File: examples/Erlang/tasksink2.erl
#! /usr/bin/env escript
%%
%% Task sink - design 2
%% Adds pub-sub flow to send kill signal to workers
%%

main(_) ->
    {ok, Context} = erlzmq:context(),

    %% Socket to receive messages on
    {ok, Receiver} = erlzmq:socket(Context, pull),
    ok = erlzmq:bind(Receiver, "tcp://*:5558"),

    %% Socket for worker control
    {ok, Controller} = erlzmq:socket(Context, pub),
    ok = erlzmq:bind(Controller, "tcp://*:5559"),

    %% Wait for start of batch
    {ok, _} = erlzmq:recv(Receiver),

    %% Start our clock now
    Start = now(),

    %% Process 100 confirmations
    process_confirmations(Receiver, 100),
    io:format("Total elapsed time: ~b msec~n",
              [timer:now_diff(now(), Start) div 1000]),

    %% Send kill signal to workers
    ok = erlzmq:send(Controller, <<"KILL">>),

    %% Finished
    ok = erlzmq:close(Controller),
    ok = erlzmq:close(Receiver),
    ok = erlzmq:term(Context, 1000).

process_confirmations(_Receiver, 0) -> ok;
process_confirmations(Receiver, N) when N > 0 ->
    {ok, _} = erlzmq:recv(Receiver),
    case N - 1 rem 10 of
        0 -> io:format(":");
        _ -> io:format(".")
    end,
    process_confirmations(Receiver, N - 1).

File: examples/Haskell/tasksink2.hs
{-# LANGUAGE OverloadedStrings #-}

--  Task sink - design 2
--  Adds pub-sub flow to send kill signal to workers

module Main where

import Control.Monad
import Data.Time.Clock
import System.IO
import System.ZMQ4.Monadic

main :: IO ()
main = runZMQ $ do
    -- Socket to receive messages on
    receiver <- socket Pull
    bind receiver "tcp://*:5558"

    -- Socket for worker control
    controller <- socket Pub
    bind controller "tcp://*:5559"

    -- Wait for start of batch
    _ <- receive receiver

    -- Start our clock now
    start_time <- liftIO getCurrentTime

    -- Process 100 confirmations
    liftIO $ hSetBuffering stdout NoBuffering
    forM_ [1..100] $ \i -> do
        _ <- receive receiver
        if i `mod` 10 == 0
            then liftIO $ putStr ":"
            else liftIO $ putStr "."

    end_time <- liftIO getCurrentTime
    liftIO . putStrLn $ "Total elapsed time: " ++ show (diffUTCTime end_time start_time * 1000) ++ " msec"

    -- Send kill signal to workers
    send controller [] "KILL"
